### ISSUE TITLE
Rails scaffold loses data when editing model with jsonb data when jsonb data is not edited and rails writes nil to db

### ISSUE THREAD

**Issue by iosdev-republicofapps** [rails/rails#26153]
### Steps to reproduce
shell> rails new jsontest --database=postgresql
shell> rails g scaffold campaign name:string cost:integer data:jsonb
shell> rails s

Now go to /campaigns, then click create to create a campaign.
Create a campaign with data like name=campaign 1, cost=100, data={"foo" : 2} and save.
The campaign will be properly saved and the jsonb data will be properly persisted in Postgres.
Now click on the edit link in the scaffold for this campaign.  Edit the cost only, say change it to 101, then click save (without editing the data (jsonb) field in any way, without even giving focus to its form field).  Once saved, you'll notice in both Postgres and rails that the campaign now has data==nil, i.e. the jsonb data field got overwritten with nil even though I never even clicked on the data form field, let alone edited it.

This is a serious data loss.  Rails scaffolds should not drop data that has not been edited.

You can see when I first create the campaign that the server log shows:

```
Started POST "/campaigns" for ::1 at 2016-08-13 14:15:33 -0700
Processing by CampaignsController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"Nhy7w1uZVKVnPk4HA36OVfuKkIszwsUR1QnaoiQKzD91zSGkw1kj4DdqKGJ4JTxoTqF0EQxctzltluQmnXQl1g==", "campaign"=>{"name"=>"campaign1", "cost"=>"100", "data"=>"{ \"foo\" : 2 }"}, "commit"=>"Create Campaign"}
   (0.1ms)  BEGIN
  SQL (0.2ms)  INSERT INTO "campaigns" ("name", "cost", "data", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["name", "campaign1"], ["cost", 100], ["data", "{\"foo\":2}"], ["created_at", "2016-08-13 21:15:33.788092"], ["updated_at", "2016-08-13 21:15:33.788092"]]
   (5.9ms)  COMMIT
Redirected to http://localhost:3000/campaigns/2
Completed 302 Found in 9ms (ActiveRecord: 6.2ms)
```

Then when I edit:
```
Started PATCH "/campaigns/2" for ::1 at 2016-08-13 14:16:45 -0700
Processing by CampaignsController#update as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"HfOQx8pXjcnOCV0lc3KJ1Sd3Oxah/pmzPjOxxF1dZ0NeIgqgUpf6jJ5dO0AIKTvoklzfjJ5g65uGrI9A5COOqg==", "campaign"=>{"name"=>"campaign1", "cost"=>"101", "data"=>"{\"foo\"=>2}"}, "commit"=>"Update Campaign", "id"=>"2"}
  Campaign Load (0.1ms)  SELECT  "campaigns".* FROM "campaigns" WHERE "campaigns"."id" = $1 LIMIT 1  [["id", 2]]
   (0.1ms)  BEGIN
  SQL (0.2ms)  UPDATE "campaigns" SET "cost" = $1, "data" = $2, "updated_at" = $3 WHERE "campaigns"."id" = $4  [["cost", 101], ["data", nil], ["updated_at", "2016-08-13 21:16:45.044172"], ["id", 2]]
   (0.7ms)  COMMIT
Redirected to http://localhost:3000/campaigns/2
Completed 302 Found in 4ms (ActiveRecord: 1.0ms)
```
Notice how the form is sending back the correct data field value (i.e. not nil) but that for some reason, ActiveRecord or something is setting data to nil.

This is wrong.  I did not edit the data field.  There is no reason for rails to set this field to nil when updating the record in the database.

(Guidelines for creating a bug report are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

### Expected behavior
Tell us what should happen

Rails should not lose data in scaffolds.  Rails should not overwrite unchanged jsonb fields with nil.

### Actual behavior
Tell us what happens instead

Rails does lose data in scaffolds.  Rails does overwrite unchanged jsonb fields with nil.

### System configuration
**Rails version**:
4.2.7.1

**Ruby version**:
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin15]

postgres 9.5.3

**To be clear:** I have not written a single line of code here: `rails new` and `rails g scaffold` are the only things that generated code here. This is all rails.  There is no "user error" involved here.

Thanks!  :-)

**Comment by sgrif** [comment]
So the issue here is basically that the scaffold generator was never intended to work with the JSON datatype. Since it doesn't recognize the type, it's assuming it should go in a string input and calling `to_s` on the value. Since `{"foo"=>2}` isn't valid JSON, the type is throwing it out (we always cast junk input to nil rather than raising).

In Rails 5, ironically, if you did the same you would be persisting the string `"{\"foo\":2}"`, as that's the JSON primitive you requested.

Basically the short version is: Yes, this doesn't work. The scaffold generator isn't meant to give you a working application for every data type, it's meant to give you a place to start. The JSON data type isn't meant to work directly from text fields like the scaffold generator gives. But this isn't a bug.
